File: mock_data.py
# Database seeder file.

# Run this ONCE after setting up the project to pre-populate
# the database with all real books from the shelves and the
# three test accounts.

# It is safe to re-run — books and users that already exist are skipped (no duplicates, no errors).

import sqlite3
import bcrypt
from pathlib import Path

DB_PATH = Path(__file__).parent / "cs_library.db"


# LOCAL BOOK CACHE
# This cache contains the real books we scanned from Shelves 1-5.

BOOKS = [
    # Shelf 1
    ("0471170828",    "Applied Regression Analysis",           "Draper & Smith",       "https://covers.openlibrary.org/b/isbn/0471170828-L.jpg",    "Available", "Shelf 1"),
    ("0517887290",    "Fingerprints of the Gods",              "Graham Hancock",        "https://covers.openlibrary.org/b/isbn/0517887290-L.jpg",    "Available", "Shelf 1"),
    # Shelf 2
    ("9780073523262", "Computer Networks: A Top Down Approach","Behrouz A. Forouzan",  "https://covers.openlibrary.org/b/isbn/9780073523262-L.jpg", "Available", "Shelf 2"),
    ("0321269675",    "Using UML",                             "Perdita Stevens",       "https://covers.openlibrary.org/b/isbn/0321269675-L.jpg",    "Available", "Shelf 2"),
    ("9781491960202", "Learning Web Design",                   "Jennifer Robbins",      "https://covers.openlibrary.org/b/isbn/9781491960202-L.jpg", "Available", "Shelf 2"),
    ("9781783982141", "Kali Linux Network Scanning",           "Justin Hutchens",       "https://covers.openlibrary.org/b/isbn/9781783982141-L.jpg", "Available", "Shelf 2"),
    ("1590593898",    "Joel on Software",                      "Joel Spolsky",          "https://covers.openlibrary.org/b/isbn/1590593898-L.jpg",    "Available", "Shelf 2"),
    # Shelf 3
    ("9781782163121", "Mastering Kali Linux",                  "Robert W. Beggs",       "https://covers.openlibrary.org/b/isbn/9781782163121-L.jpg", "Available", "Shelf 3"),
    ("0471098779",    "Elements of Cartography",               "Robinson et al.",       "https://covers.openlibrary.org/b/isbn/0471098779-L.jpg",    "Available", "Shelf 3"),
    ("0131478230",    "Practical Guide to Linux",              "Mark G. Sobell",        "https://covers.openlibrary.org/b/isbn/0131478230-L.jpg",    "Available", "Shelf 3"),
    ("076372677X",    "Information Security Illuminated",      "Solomon & Chapple",     "https://covers.openlibrary.org/b/isbn/076372677X-L.jpg",    "Available", "Shelf 3"),
    ("0596005458",    "Security Warrior",                      "Peikari & Chuvakin",    "https://covers.openlibrary.org/b/isbn/0596005458-L.jpg",    "Available", "Shelf 3"),
    # Shelf 4
    ("0030977177",    "World Regional Geography",              "Wheeler & Kostbade",    "https://covers.openlibrary.org/b/id/6463996-L.jpg",         "Available", "Shelf 4"),
    ("0314004246",    "Physical Geography Manual",             "West Publishing",       "https://covers.openlibrary.org/b/id/11145828-L.jpg",        "Available", "Shelf 4"),
    # Shelf 5
    ("0671492071",    "Alan Turing: The Enigma",               "Andrew Hodges",         "https://covers.openlibrary.org/b/isbn/0671492071-L.jpg",    "Available", "Shelf 5"),
    ("0321534964",    "The Art of Computer Programming",       "Donald Knuth",          "https://covers.openlibrary.org/b/isbn/0321534964-L.jpg",    "Available", "Shelf 5"),
]


# TEST ACCOUNTS
# NEW: Added 'id' key inside each user dict so current_user['id'] works in main.py (US006)

TEST_USERS = [
    (12345, "Kenneth Molina",  "molinak4@southernct.edu",    "changeme123"),
    (11111, "Jose Gaspar",     "gasparmarij1@southernct.edu","changeme123"),
    (99999, "Professor James", "james@southernct.edu",        "changeme123"),
]


# SEEDING FUNCTION
def seed():
    if not DB_PATH.exists():
        print("X  cs_library.db not found.")
        print("   Run main.py first so init_db() creates it, then re-run this script.")
        return

    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA foreign_keys = ON")

    # Ensure shelf column exists (safe migration)
    cols = [row[1] for row in conn.execute("PRAGMA table_info(books)").fetchall()]
    if "shelf" not in cols:
        conn.execute("ALTER TABLE books ADD COLUMN shelf TEXT NOT NULL DEFAULT ''")
        conn.commit()
        print("  Migrated: added 'shelf' column to books table.")

    # Seed books
    books_added = 0
    books_skipped = 0
    for isbn, title, author, cover, status, shelf in BOOKS:
        existing = conn.execute(
            "SELECT isbn FROM books WHERE isbn = ?", (isbn,)
        ).fetchone()
        if existing:
            books_skipped += 1
            continue
        conn.execute(
            "INSERT INTO books (isbn, title, author, cover, status, shelf) "
            "VALUES (?, ?, ?, ?, ?, ?)",
            (isbn, title, author, cover, status, shelf),
        )
        books_added += 1

    conn.commit()
    print(f"  Books: {books_added} added, {books_skipped} already existed -- skipped.")

    # Seed test users with explicit IDs
    users_added = 0
    users_skipped = 0
    for user_id, name, email, password in TEST_USERS:
        # Skip if this ID or email already exists
        existing = conn.execute(
            "SELECT id FROM users WHERE student_id = ? OR email = ? COLLATE NOCASE",
            (user_id, email),
        ).fetchone()
        if existing:
            users_skipped += 1
            continue

        pw_hash = bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode()
        conn.execute(
            "INSERT INTO users (student_id, name, email, password_hash, active) "
            "VALUES (?, ?, ?, ?, 1)",
            (str(user_id), name, email.lower(), pw_hash),
        )
        users_added += 1

    conn.commit()
    conn.close()

    print(f"  Users: {users_added} added, {users_skipped} already existed -- skipped.")

    if users_added > 0:
        print()
        print("  Test account credentials:")
        print(f"  {'ID':<8} {'Name':<22} {'Email':<35} Password")
        print(f"  {'-'*8} {'-'*22} {'-'*35} {'-'*12}")
        for user_id, name, email, password in TEST_USERS:
            print(f"  {user_id:<8} {name:<22} {email:<35} {password}")

    print()
    print("Seeding complete.")


if __name__ == "__main__":
    seed()

File: src/mock-data.ts
// Database seeder file.
// Run this once after setting up the project to populate
// the database with books and test accounts.
//
// Run the following commands in order:
//
// npm run db:generate
// npm run db:migrate
// npx tsx src/mock-data.ts
//
// Safe to re-run — duplicates are skipped.

import { drizzle } from 'drizzle-orm/node-postgres';
import { Pool } from 'pg';
import { eq, or } from 'drizzle-orm';
import bcrypt from 'bcryptjs';
import * as dotenv from 'dotenv';

dotenv.config();

import { users, books } from './db/schema/schema.js';

// Connect to the database

const pool = new Pool({ connectionString: process.env['DATABASE_URL'] });
const db   = drizzle(pool, { schema: { users, books } });

// Pre-existing books

const BOOKS = [
  // Shelf 1
  { isbn: "0471170828",    title: "Applied Regression Analysis",            author: "Draper & Smith",        cover: "https://covers.openlibrary.org/b/isbn/0471170828-L.jpg",    status: "Available", shelf: "Shelf 1" },
  { isbn: "0517887290",    title: "Fingerprints of the Gods",               author: "Graham Hancock",         cover: "https://covers.openlibrary.org/b/isbn/0517887290-L.jpg",    status: "Available", shelf: "Shelf 1" },
  // Shelf 2
  { isbn: "9780073523262", title: "Computer Networks: A Top Down Approach", author: "Behrouz A. Forouzan",   cover: "https://covers.openlibrary.org/b/isbn/9780073523262-L.jpg",  status: "Available", shelf: "Shelf 2" },
  { isbn: "0321269675",    title: "Using UML",                              author: "Perdita Stevens",        cover: "https://covers.openlibrary.org/b/isbn/0321269675-L.jpg",    status: "Available", shelf: "Shelf 2" },
  { isbn: "9781491960202", title: "Learning Web Design",                    author: "Jennifer Robbins",       cover: "https://covers.openlibrary.org/b/isbn/9781491960202-L.jpg",  status: "Available", shelf: "Shelf 2" },
  { isbn: "9781783982141", title: "Kali Linux Network Scanning",            author: "Justin Hutchens",        cover: "https://covers.openlibrary.org/b/isbn/9781783982141-L.jpg",  status: "Available", shelf: "Shelf 2" },
  { isbn: "1590593898",    title: "Joel on Software",                       author: "Joel Spolsky",           cover: "https://covers.openlibrary.org/b/isbn/1590593898-L.jpg",    status: "Available", shelf: "Shelf 2" },
  // Shelf 3
  { isbn: "9781782163121", title: "Mastering Kali Linux",                   author: "Robert W. Beggs",        cover: "https://covers.openlibrary.org/b/isbn/9781782163121-L.jpg",  status: "Available", shelf: "Shelf 3" },
  { isbn: "0471098779",    title: "Elements of Cartography",                author: "Robinson et al.",        cover: "https://covers.openlibrary.org/b/isbn/0471098779-L.jpg",    status: "Available", shelf: "Shelf 3" },
  { isbn: "0131478230",    title: "Practical Guide to Linux",               author: "Mark G. Sobell",         cover: "https://covers.openlibrary.org/b/isbn/0131478230-L.jpg",    status: "Available", shelf: "Shelf 3" },
  { isbn: "076372677X",    title: "Information Security Illuminated",       author: "Solomon & Chapple",      cover: "https://covers.openlibrary.org/b/isbn/076372677X-L.jpg",    status: "Available", shelf: "Shelf 3" },
  { isbn: "0596005458",    title: "Security Warrior",                       author: "Peikari & Chuvakin",     cover: "https://covers.openlibrary.org/b/isbn/0596005458-L.jpg",    status: "Available", shelf: "Shelf 3" },
  // Shelf 4
  { isbn: "0030977177",    title: "World Regional Geography",               author: "Wheeler & Kostbade",     cover: "https://covers.openlibrary.org/b/id/6463996-L.jpg",         status: "Available", shelf: "Shelf 4" },
  { isbn: "0314004246",    title: "Physical Geography Manual",              author: "West Publishing",        cover: "https://covers.openlibrary.org/b/id/11145828-L.jpg",        status: "Available", shelf: "Shelf 4" },
  // Shelf 5
  { isbn: "0671492071",    title: "Alan Turing: The Enigma",                author: "Andrew Hodges",          cover: "https://covers.openlibrary.org/b/isbn/0671492071-L.jpg",    status: "Available", shelf: "Shelf 5" },
  { isbn: "0321534964",    title: "The Art of Computer Programming",        author: "Donald Knuth",           cover: "https://covers.openlibrary.org/b/isbn/0321534964-L.jpg",    status: "Available", shelf: "Shelf 5" },
] as const;

// Test Accounts
// student_id stored as a string to match the VARCHAR column.
// Passwords are hashed by this script — never stored in plain text.
// Although unnecessary since these are test accounts, change passwords before deploying to production!

const TEST_USERS = [
  { studentId: "40000", name: "Nisreen Cain",     email: "cainn3@southernct.edu", password: "changeme123" },
  { studentId: "10101", name: "Amilcar Armmand",  email: "armmanda1@southernct.edu",    password: "changeme123" },
  { studentId: "12345", name: "Kenneth Molina",  email: "molinak4@southernct.edu",    password: "changeme123" },
  { studentId: "11111", name: "Jose Gaspar",     email: "gasparmarij1@southernct.edu", password: "changeme123" },
  { studentId: "99999", name: "Professor James", email: "james@southernct.edu",         password: "changeme123" },
];

// Seed Function

async function seed() {
  console.log('Starting database seed...\n');

  // Book Provisioning

  let booksAdded   = 0;
  let booksSkipped = 0;

  for (const book of BOOKS) {
    const [existing] = await db
      .select({ isbn: books.isbn })
      .from(books)
      .where(eq(books.isbn, book.isbn))
      .limit(1);

    if (existing) {
      booksSkipped++;
      continue;
    }

    await db.insert(books).values(book);
    booksAdded++;
  }

  console.log(`  Books : ${booksAdded} added, ${booksSkipped} already existed — skipped.`);

  // User provisioning

  let usersAdded   = 0;
  let usersSkipped = 0;

  for (const { studentId, name, email, password } of TEST_USERS) {
    const [existing] = await db
      .select({ id: users.id })
      .from(users)
      .where(
        or(
          eq(users.studentId, studentId),
          eq(users.email, email.toLowerCase()),
        )
      )
      .limit(1);

    if (existing) {
      usersSkipped++;
      continue;
    }

    const passwordHash = await bcrypt.hash(password, 12);

    await db.insert(users).values({
      studentId,
      name,
      email:           email.toLowerCase(),
      passwordHash,
      active:          true,
    });

    usersAdded++;
  }

  console.log(`  Users : ${usersAdded} added, ${usersSkipped} already existed — skipped.`);

  // Output the summary.

  if (usersAdded > 0) {
    console.log('\n  Test account credentials:');
    console.log(`  ${'Student ID'.padEnd(12)} ${'Name'.padEnd(22)} ${'Email'.padEnd(35)} Password`);
    console.log(`  ${'-'.repeat(12)} ${'-'.repeat(22)} ${'-'.repeat(35)} ${'-'.repeat(12)}`);
    for (const { studentId, name, email, password } of TEST_USERS) {
      console.log(`  ${studentId.padEnd(12)} ${name.padEnd(22)} ${email.padEnd(35)} ${password}`);
    }
  }

  console.log('\nSeeding complete.\n');
  await pool.end();
}

seed().catch((err) => {
  console.error('SEEDING FAILED:', err);
  pool.end();
  process.exit(1);
});